src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+#include <Corrade/Utility/ConfigurationGroup.h>
+#include <Corrade/Utility/Math.h>
+#include <Magnum/Math/Matrix3.h>
+#include <Magnum/Math/RectangularMatrix.h>
+#include <Magnum/Math/Vector3.h>
+#include <Magnum/Math/Quaternion.h>
+#include <Magnum/Trade/MeshData.h>
+#include <Magnum/Trade/CameraData.h>
+#include <Magnum/Trade/MaterialData.h>
+#include <Magnum/Trade/TextureData.h>
+#include <Magnum/Trade/ImageData.h>
+#include <Magnum/Trade/LightData.h>
+#include <Magnum/Trade/SceneData.h>
+#include <MagnumPlugins/AnyImageImporter/AnyImageImporter.h>
+
+namespace Magnum { namespace Trade {
+    
+using namespace Containers::Literals;
+
+namespace {
+
+struct MeshChunk {
+    uint32_t meshId;
+    uint32_t meshMaterialIndex;
+};
+
+ufbx_load_opts loadOptsFromConfiguration(Utility::ConfigurationGroup& conf) {
+    static_cast<void>(conf);
+
+    ufbx_load_opts opts = { };
+
+    opts.generate_missing_normals = conf.value<bool>("generateMissingNormals");
+    opts.strict = conf.value<bool>("strict");
+    opts.disable_quirks = conf.value<bool>("disableQuirks");
+    opts.load_external_files = conf.value<bool>("loadExternalFiles");
+    opts.ignore_geometry = conf.value<bool>("ignoreGeometry");
+    opts.ignore_animation = conf.value<bool>("ignoreAnimation");
+    opts.ignore_embedded = conf.value<bool>("ignoreEmbedded");
+    opts.ignore_all_content = conf.value<bool>("ignoreAllContent");
+
+    /* We need to split meshes by material so create a dummy ufbx_mesh_material
+       containing the whole mesh to make processing code simpler. */
+    opts.allow_null_material = true;
+
+    return opts;
+}
+
+inline Int typedId(ufbx_element *element) {
+    return element ? (Int)element->typed_id : -1;
+}
+
+inline SamplerWrapping toSamplerWrapping(ufbx_wrap_mode mode) {
+    if (mode == UFBX_WRAP_CLAMP) {
+        return SamplerWrapping::ClampToEdge;
+    } else if (mode == UFBX_WRAP_REPEAT) {
+        return SamplerWrapping::Repeat;
+    } else {
+        /* @todo: What to do about unhandled enums */
+        return SamplerWrapping::Repeat; /* LCOV_EXCL_LINE */
+    }
+}
+
+inline Containers::String asString(ufbx_string str) {
+    return Containers::String{str.data, str.length};
+}
+
+inline Containers::StringView asStringView(ufbx_string str) {
+    return Containers::StringView{str.data, str.length, Containers::StringViewFlag::NullTerminated};
+}
+
+inline Vector2 asVector2(const ufbx_vec2 &v) {
+    return Vector2{ (Float)v.x, (Float)v.y };
+}
+
+inline Vector3 asVector3(const ufbx_vec3 &v) {
+    return Vector3{ (Float)v.x, (Float)v.y, (Float)v.z };
+}
+
+inline Vector4 asVector4(const ufbx_vec4 &v) {
+    return Vector4{ (Float)v.x, (Float)v.y, (Float)v.z, (Float)v.w };
+}
+
+inline Color4 asColor4(const ufbx_vec4 &v) {
+    return Color4{ (Float)v.x, (Float)v.y, (Float)v.z, (Float)v.w };
+}
+
+inline Quaterniond asQuaterniond(const ufbx_quat &q) {
+    return Quaterniond{ Vector3d{q.x, q.y, q.z}, q.w };
+}
```

Comment:
FYI, all math and string classes have a possibility to define conversion constructors for external types:

```cpp
namespace Corrade { namespace Containers { namespace Implementation {

template<> struct StringConverter<const char, ufbx_string> {
    static String from(ufbx_string other) {
        return String{other.data, other.length};
    }
};
template<> struct StringViewConverter<const char, ufbx_string> {
    static StringView from(ufbx_string other) {
        return StringView{other.data, other.length, Containers::StringViewFlag::NullTerminated};
    }
};

}}}

namespace Magnum { namespace Math { namespace Implementation {

template<> struct VectorConverter<3, Float, ufbx_vec3> {
    constexpr static Vector<3, Float> from(const ufbx_vec3& other) {
        return {Float(other.x), Float(other.y), Float(other.z)};
    }
};

// probably also this?
template<> struct VectorConverter<3, Double, ufbx_vec3> {
    constexpr static Vector<3, Double> from(const ufbx_vec3& other) {
        return {other.x, other.y, other.z};
    }
};

}}}
```

Similarly for vectors of other sizes, matrices (`RectangularMatrixConverter<cols, rows, T, ExternalType>`, such as for `uv_to_texture`) and quats (`QuaternionConverter<T, ExternalType>`). 

Then in the code it's an implicit conversion for strings and *explicit* conversion for math types, so e.g.:

```cpp
ufbx_string a;
Containers::StringView b = a;
...
translations[nodeId] = Vector3d(node->local_transform.translation);
```

Reply:
Ah that looks a lot cleaner, I'll do that! I guess it's still safe to use those implementation details in "user" code without them breaking in future versions?

Reply:
Ah, sorry I should have clarified the meaning of "implementation detail" -- it's useful for library / plugin implementers, but not for end users, which is why no public docs. But the interface is stable, and various "integration" libraries like [GlmIntegration](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1GlmIntegration.html) or [EigenIntegration](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1EigenIntegration.html) rely on these, so it's not like it'd be peeking into some fragile internals :)